lunabot_embedded/include/lunabot_embedded/sensor_proc.h

```diff
@@ -0,0 +1,68 @@
+#include <algorithm>
+#include <math.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <termios.h>
+#include <array>
+
+using namespace std;
+
+#define DEG2RAD(x) ((x)*M_PI / 180.0)
+
+template <size_t N>
+struct RingBuffer {
+
+  void push(float value) {
+    sum -= data[head];
+    sum += value;
+    data[head] = value;
+    head = (head + 1) % data.size();
+  }
+
+  float mean() { return sum / data.size(); }
+
+  std::array<float,N> data;
+  int head;
+  float sum;
+};
+
+float u_mod(float n, float base) {
+  float m = fmod(n, base);
+  return (m >= 0) ? m : m + base;
+}
+
+// TODO: Fix cyclic spikes in delta
+float deg_angle_delta(float deg, float prev_deg) {
+  float raw = deg - prev_deg;
+  if (raw == 0) {
+    return 0;
+  }
+  float turn = min(u_mod(raw, 360.), u_mod(-raw, 360.));
+  int dir = abs(raw) / raw;
+  return turn * dir;
+}
+
+// alpha = 0, no filter, alpha = 1, full
+float leaky_integrator(float value, float prev_filter_value, float alpha) {
+  return alpha * prev_filter_value + (1 - alpha) * value;
+}
+
+void angle_noise_rej_filter(float *curr_angle_deg, float *prev_angle_deg, float max_delta) {
```

Comment:
Can you explain what the else condition is for? I'm a bit confused by that.

Reply:
If the else condition runs, it sets the previous valid angle to the current valid angle. where validity is if the angle was NOT rejected.